Fix return type Any for Ternary ops

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49165 Fix return type Any for Ternary ops**

Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

ejguan committed:
Fix return type Any for Ternary ops

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"




[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"




[ghstack-poisoned]

ejguan commented:
Hey @eellison, I am wondering if and where I need to add tests to this behavior?

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1250,11 +1250,14 @@ struct to_ir {
+      const TernaryIf& expr,
+      const TypePtr& type_hint = nullptr) {
     CondValue cond_value = emitCondExpr(expr.cond());
     auto true_expr = [&] { return emitExpr(expr.true_expr()); };
```
Can we add the type hint into the `emitExpr` of `true_expr` and `false_expr` ? I think that would make it so we don't need to touch `emitListComprehension` 

Can you add a test like `Optional[int] = None if cond else 1` to show that logic works not just for `Any` ?

ejguan replied:
Added type hint into `emitExpr`, and moved `unCheckedCast` to it.

eellison reviewed (commented):
Nice! Good start. One small change and i think we're good to go.

You can add the test next to herre:  https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L6532

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

eellison commented on test/test_jit.py:
```diff
@@ -6490,6 +6490,24 @@ def func(a, b):
+            def forward(self) -> Optional[int]:
+                return None if self.training else 1
+
+        self.checkModule(M1(), ())
```
maybe worth checking the module with M1.training = false just for completeness

ejguan replied:
Will do

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -2954,7 +2956,12 @@ struct to_ir {
-    return emitSugaredExpr(tree, 1, type_hint)->asValue(tree.range(), method);
+    Value* out_val =
+        emitSugaredExpr(tree, 1, type_hint)->asValue(tree.range(), method);
+    if (type_hint == AnyType::get() && out_val->type() != AnyType::get()) {
```
Do all the tests work if you remove this? If yes, could you remove this ? I 

If no, could you rewrite this to add an unchecked cast if `*out_val->type() != *type_hint && out_val->isSubtypeOf(type_hint)` ?

ejguan replied:
It doesn't work if I remove it, since `emitSugaredExpr` will ignore the `type_hint` for the case `TK_VAR `.
And, I don't think it should be changed to `subtype` for the output. (it will raise error for my test case `M3`)
```python
class M3(torch.nn.Module):
    def forward(self) -> Optional[int]:
        return None if self.training else 1
```
It's not reasonable to cast the output value from `int` to `Optional[int]` or `None` to `Optional[int]`.
The `unchecked_cast` should be used for `Any`.

ejguan replied:
I will verify it using 
-> Union[int, str]
1 if self.training else 'xxx'

ejguan replied:
Interesting, `Union` is not supported in JIT. 
So, IMO we should not to cast for all subtype. In JIT, reasonable subtypes only happen for `Optional` (no cast) and `Any` (required cast)

eellison replied:
So, the reason we need to add the cast is because by default we do not unify two types to Any.
```
if cond:
    a = 1.
else 
    a = 1
```
I think this is reasonable - Any is hard  to use, and in most cases you want to  fix the code there so  that both branches are floats. There's actually one other case where we have a type which is a SuperType which we don't by default unify to - [NumberType](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/type.cpp#L249). This is an internal type though, I don't think anyone uses it. The `*out_val->type() != *type_hint && out_val->isSubtypeOf(type_hint)` is generically valid, but it's true that it only applies to `Any` right now. 

Could you add the following comment?
> AnyType is the only user-exposed type which we don't unify to from its subtypes, so we add a cast for use cases like `x : Any = 1 if cond else "str"`

ejguan replied:
Thanks @eellison . Will add the comment and ship.

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

eellison reviewed (approved):
Cool, great job!  Please add a comment & then ship it 🚢

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]

ejguan committed:
Update on "Fix return type Any for Ternary ops"


Fixes #43942
Differential Revision: [D25463694](https://our.internmc.facebook.com/intern/diff/D25463694)

```python
>>> class M2(torch.nn.Module):
...   def forward(self) -> Any:
...         return 'xx' if self.training else {}
>>> s2 = torch.jit.script(M2())
>>> print(s2.graph)
graph(%self : __torch__.M2):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:13:15
  %1 : bool = prim::GetAttr[name="training"](%self)
  %6 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:13:15
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%6)
```
```python
>>> class M3(torch.nn.Module):
...     def forward(self) -> Any:
...         out: Any = 'xx' if self.training else {}
...         return out
>>> s3 = torch.jit.script(M3())
>>> print(s3.graph)
graph(%self : __torch__.M3):
  %2 : str = prim::Constant[value="xx"]() # /Users/erjia/workspace/test/jit/jit.py:17:19
  %1 : bool = prim::GetAttr[name="training"](%self)
  %out.1 : Any = prim::If(%1) # /Users/erjia/workspace/test/jit/jit.py:17:19
    block0():
      %3 : Any = prim::unchecked_cast(%2)
      -> (%3)
    block1():
      %4 : Dict(str, Tensor) = prim::DictConstruct()
      %5 : Any = prim::unchecked_cast(%4)
      -> (%5)
  return (%out.1)
```

[ghstack-poisoned]
